server/main_server/src/rodsServer.cpp

```diff
@@ -434,36 +455,396 @@ namespace
+            }
+            log::trace("Local server's hostname = [{}]", hostname);
+
+            std::string leader;
+            std::string successor;
+
+            {
+                ix::client_connection comm;
+
+                auto result = get_grid_configuration_option_value(comm, "delay_server", "leader");
+                if (!result) {
+                    return;
+                }
+
+                leader = std::move(*result);
+                log::trace("Leader's hostname = [{}]", leader);
+
+                result = get_grid_configuration_option_value(comm, "delay_server", "successor");
+                if (!result) {
+                    return;
+                }
+
+                successor = std::move(*result);
+                log::trace("Successor's hostname = [{}]", successor);
+            }
+
+            if (hostname == leader) {
+                log::trace("Local server [{}] is the leader.", hostname);
+
+                if (!successor.empty() && hostname != successor) {
+                    log::trace("New successor detected!");
+
+                    const auto pid = irods::get_delay_server_pid();
+
+                    if (!pid) {
+                        log::trace("Could not get delay server PID.");
+                        return;
+                    }
+
+                    log::trace("Delay server PID = [{}]", *pid);
+                    log::info("Sending shutdown signal to delay server.");
+
+                    // If the delay server is running locally, send SIGTERM to it and wait
+                    // for graceful shutdown. The call to waitpid() will cause the CRON manager
+                    // to stop making process until it returns because it is single threaded.
+                    kill(*pid, SIGTERM);
+
+                    int child_status = 0;
+                    waitpid(*pid, &child_status, 0);
+                    log::info("Delay server has completed shutdown [exit_code={}].", WEXITSTATUS(child_status));
+                }
+                else {
+                    log::trace("No successor detected. Checking if delay server needs to be launched ...");
+
+                    // If the delay server is not running, launch it!
+                    if (const auto pid = irods::get_delay_server_pid(); !pid) {
+                        launch_delay_server(_enable_test_mode, _write_to_stdout);
+                    }
+                }
+            }
+            else if (hostname == successor) {
+                log::trace("Local server [{}] is the successor.", hostname);
+
+                try {
+                    if (const auto res = is_delay_server_running_on_remote_host(leader); res && !*res) {
+                        {
+                            ix::client_connection comm;
+                            set_grid_configuration_option_value(comm, "delay_server", "leader", hostname);
+                            set_grid_configuration_option_value(comm, "delay_server", "successor", "");
```

Comment:
these should probably be updated together atomically.

update and use atomic_apply_database_operations if we have time, but not urgent.

Reply:
done